intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c

```diff
@@ -7946,6 +7955,50 @@ static void ne_compute_forward_rope_f16(const struct ne_compute_params* params,
   }
 }
 
+static void ne_compute_forward_rope_jblas(const struct ne_compute_params* params, const struct ne_tensor* src0,
+                                          const struct ne_tensor* src1, struct ne_tensor* dst) {
+  if (params->type == NE_TASK_INIT || params->type == NE_TASK_FINALIZE) return;
+  NE_ASSERT(("use internal multi-threading", params->nth == 1));
+
+  NE_ASSERT(src1->type == NE_TYPE_I32);
+  NE_ASSERT(ne_nelements(src1) == 5);  // 5 params
+
+  const int n_past = ((int32_t*)src1->data)[0];
+  const int n_dims = ((int32_t*)src1->data)[1];
+  const int mode = ((int32_t*)src1->data)[2];
+  const int prompt_size = ((int32_t*)src1->data)[3];
+  const int n_keep = ((int32_t*)src1->data)[4];
+  const bool skip = mode & 1;
+  const bool is_neox = mode & 2;
+  const bool is_glm = mode & 4;
+  NE_ASSERT(("glm mode RoPE is not implemented!", !is_glm));
+  const bool is_shift = n_keep >= 0;
+  NE_ASSERT(("shift RoPE is only implemented for the vanilla mode", !is_shift || !(is_glm || is_neox || skip)));
+
+  const int batch_size = dst->ne[3];
+  const int head_num = dst->ne[2];
+  const int seq_len = dst->ne[1];
+  const int head_size = dst->ne[0];
+
+  if (is_shift) {
+    ne_fp16_t* cossin = (dst->opt[0] != NULL) ? dst->opt[0]->data : NULL;
+    if (cossin == NULL) {
+      float theta = n_past;
+      const float theta_scale = powf(10000.0, -2.0f / n_dims);
```

Comment:
I think ```n_dims``` and the related values are static, right?

Reply:
I think is it better to leave it as it is, just like the implementation of f32/f16 rope. `n_dims` typically equals to the head size and it is passed in as a parameter.